utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.lwdataprovider
+
+import com.exactpro.th2.common.grpc.Direction
+import com.exactpro.th2.common.message.toJson
+import com.exactpro.th2.common.message.toTimestamp
+import com.exactpro.th2.dataprovider.lw.grpc.DataProviderService
+import com.exactpro.th2.dataprovider.lw.grpc.MessageGroupsSearchRequest
+import com.exactpro.th2.dataprovider.lw.grpc.MessageSearchResponse
+import com.exactpro.th2.dataprovider.lw.grpc.MessageStream
+import com.exactpro.th2.dataprovider.lw.grpc.TimeRelation
+import io.grpc.Context
+import mu.KotlinLogging
+import java.time.Duration
+import java.time.Instant
+import java.util.function.Function
+
+/**
+ * This class provide ability to search messages via lw-data-provider gRPC service
+ * @param requestInterval interval for sending requests via [service]. Searcher uses this value to split search interval to smaller pieces to reduce volume of requested data via one time.
+ */
+class MessageSearcher @JvmOverloads constructor(
+    private val service: DataProviderService,
+    private val requestInterval: Duration = DEFAULT_REQUEST_INTERVAL,
+    private val defaultBook: String = "",
+    private val defaultSessionGroup: String? = null,
+    private val defaultSessionAlias: String = "",
+) {
+    /**
+     * Searches the first message in [BASE64_FORMAT] format matched by filter. Execute search from [Instant.now] to previous time
+     * @param sessionGroup is parameter for request. Searcher uses [sessionAlias] value when [sessionGroup] is null or blank
+     * @param interval for searching. Searcher stops to send requests when the next `from` time is before than search start time minus [interval]
+     */
+    fun searchLastOrNull(
+        book: String = defaultBook,
+        sessionGroup: String? = defaultSessionGroup,
+        sessionAlias: String = defaultSessionAlias,
```

Comment:
When you define the method with default parameters you should keep them at the end of the method. Otherwise, it is not possible to avoid them without using named arguments. Also, I think it would be good to add `@JvmOverloads` annotation to have overloads in java.
NOTE: the filter should be left at the end. Otherwise, the handy "last lambda" won't work in Kotlin